### PR TITLE
Slow and no longer existing readers slow association

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

A slow or no longer existing reader will cause the heartbeat period to
grow to the maximum.  This will persist until the defunct reader is
purged from the sytem.

Solution
--------

1. Remove the logic for computing an adaptive heartbeat period.

2. Use a separate timer for nack responses.

3. Add filtering based on reflecting heartbeat counts that discards
   acknacks that are too old.  This is to handle cases where the
   round-trip latency is greater than the heartbeat period.

### CODE REVIEW

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 1]
```diff
@@ -3032,11 +3037,12 @@ void RtpsUdpDataLink::RtpsWriter::process_nackfrag(const RTPS::NackFragSubmessag
...
   reader->requested_frags_[seq] = nackfrag.fragmentNumberState;
   readers_expecting_data_.insert(reader);
-  heartbeat_.schedule(heartbeat_period_);
+  readers_expecting_heartbeat_.insert(reader);
+  nack_response_.schedule(link->config().nak_response_delay_);
 }
 
 void
-RtpsUdpDataLink::RtpsWriter::send_and_gather_nack_replies_i(MetaSubmessageVec& meta_submessages)
+RtpsUdpDataLink::RtpsWriter::gather_nack_replies_i(MetaSubmessageVec& meta_submessages)
```
Comment (iguessthislldo) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:3045]: Please update function name in the logging.

File: tests/DCPS/SetQosPartition/Writer.cpp [hunk 2]
```diff
@@ -95,6 +95,16 @@ Writer::svc ()
...
       ACE_OS::sleep (ACE_Time_Value (0, 500000));
     }
+
+    DDS::Duration_t d = { DDS::DURATION_INFINITE_SEC, DDS::DURATION_INFINITE_NSEC };
+    ::DDS::ReturnCode_t const ret = writer_->wait_for_acknowledgments(d);
+    if (ret != DDS::RETCODE_OK)
+      {
+        ACE_ERROR ((LM_ERROR,
+                    ACE_TEXT("(%P|%t) ERROR: Writer::svc, ")
+                    ACE_TEXT ("wait_for_acknowledgments returned %d.\n"),
```
Comment (iguessthislldo) [inline tests/DCPS/SetQosPartition/Writer.cpp:105]:
```suggestion
                    ACE_TEXT("wait_for_acknowledgments returned %d.\n"),
```

Or put a space in the other one if we're sticking with the old style, I don't know, it just bothers me that they're different.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 3]
```diff
@@ -1383,6 +1367,24 @@ RtpsUdpDataLink::RtpsWriter::send_heartbeats(const DCPS::MonotonicTimePoint& now
   }
 }
 
+void
+RtpsUdpDataLink::RtpsWriter::send_nack_responses(const DCPS::MonotonicTimePoint& /*now*/)
```
Comment (iguessthislldo) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:1371]:
```suggestion
RtpsUdpDataLink::RtpsWriter::send_nack_responses(const MonotonicTimePoint& /*now*/)
```

Pretty sure this is the `DCPS` namespace.

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

**Review by iguessthislldo — APPROVED** [review]

**Comment by mitza-oci** [comment]
Parsing the testing logs on GHPRB failed.  The actual test failures were:
- tests/DCPS/Thrasher/run_test.pl high rtps
- tests/DCPS/Thrasher/run_test.pl aggressive rtps
- tests/DCPS/Thrasher/run_test.pl high rtps durable
- tests/DCPS/Thrasher/run_test.pl aggressive rtps durable
- performance-tests/bench/run_test.pl disco show-logs

**Comment by jrw972** [comment]
retest this please

**Pull request merged**